clippy_lints/src/unnecessary_self_imports.rs

```diff
@@ -0,0 +1,71 @@
+use clippy_utils::diagnostics::span_lint_and_then;
+use clippy_utils::source::snippet;
+use if_chain::if_chain;
+use rustc_ast::{Item, ItemKind, Path, PathSegment, UseTree, UseTreeKind};
+use rustc_errors::Applicability;
+use rustc_lint::{EarlyContext, EarlyLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::symbol::kw::SelfLower;
+
+declare_clippy_lint! {
+    /// **What it does:** Checks for imports ending in `::{self};`.
+    ///
+    /// **Why is this bad?** In most cases, this can be written much more cleanly by omitting `self`.
+    ///
+    /// **Known problems:** Sometimes removing `::{self}` will break code (https://github.com/rust-lang/rustfmt/issues/3568).
+    ///
+    /// **Example:**
+    ///
+    /// ```rust
+    /// use std::io::{self};
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// use std::io;
+    /// ```
+    pub UNNECESSARY_SELF_IMPORTS,
+    restriction,
+    "imports ending in `self`, which can be omitted"
+}
+
+declare_lint_pass!(UnnecessarySelfImports => [UNNECESSARY_SELF_IMPORTS]);
+
+impl EarlyLintPass for UnnecessarySelfImports {
+    fn check_item(&mut self, cx: &EarlyContext<'_>, item: &Item) {
+        if_chain! {
+            if let ItemKind::Use(
+                UseTree { kind: UseTreeKind::Nested(nodes), prefix: Path { span, .. }, .. }
+            ) = &item.kind;
+            if nodes.len() == 1;
+            if let (UseTree { prefix, kind, .. }, _) = &nodes[0];
+            if let PathSegment { ident, .. } = prefix.segments[0];
+            if ident.name == SelfLower;
+
+            then {
+                let adjusted_span = item.span.with_hi(span.hi());
+                let snippet = if let UseTreeKind::Simple(Some(alias), ..) = kind {
+                    format!(
+                        "{} as {};",
+                        snippet(cx, adjusted_span, ".."),
+                        snippet(cx, alias.span, "..")
```

Comment:
I don't think any `snippet` calls are needed.

```suggestion
                        use_tree.ident(),
                        alias,
```

Reply:
When I add `use_tree.ident()`, the program panics:
```
thread 'rustc' panicked at '`UseTree::ident` can only be used on a simple import', compiler/rustc_ast/src/ast.rs:2345:18
```
I can remove the `snippet` call for the alias, but it looks like I have to keep it in for the path.

Reply:
How about `use_tree.prefix.segments.last()?.ident`?

Reply:
That works! I have to make the span a little wonky so that it only covers the last segment in the path plus the `::{self}`, but I'm assuming that's alright?